Fix paramiko as connection via API, exception encountered: TypeError: sequence item 0

veereshkhanorkar opened the pull request:
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

connection/paramiko_ssh.py
##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```
ansible 2.1.1.0
  config file =
  configured module search path = Default w/o overrides
```
##### SUMMARY

```
While using paramiko as connection type, exception occurs:
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 124, in run
res = self.execute()
File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 448, in execute
result = self._handler.run(task_vars=variables)
File "/usr/lib/python2.7/site-packages/ansible/plugins/action/normal.py", line 33, in run
results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars))
File "/usr/lib/python2.7/site-packages/ansible/plugins/action/__init.py", line 591, in execute_module
tmp = self.make_tmp_path(remote_user)
File "/usr/lib/python2.7/site-packages/ansible/plugins/action/__init.py", line 218, in make_tmp_path
result = self.low_level_execute_command(cmd, sudoable=False)
File "/usr/lib/python2.7/site-packages/ansible/plugins/action/__init.py", line 718, in low_level_execute_command
rc, stdout, stderr = self.connection.exec_command(cmd, in_data=in_data, sudoable=sudoable)
File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/paramiko_ssh.py", line 253, in exec_command
super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/__init.py", line 51, in wrapped
self._connect()
File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/paramiko_ssh.py", line 139, in _connect
self.ssh = SSH_CONNECTION_CACHE[cache_key] = self._connect_uncached()
File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/paramiko_ssh.py", line 212, in _connect_uncached
sock_kwarg = self._parse_proxy_command(port)
File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/paramiko_ssh.py", line 153, in _parse_proxy_command
args = self._split_ssh_args(' '.join(ssh_args))
TypeError: sequence item 0: expected string, NoneType found
```

fixes #17330

Commit:
Fix for https://github.com/ansible/ansible/issues/17330

abadger commented:
Bug report is: https://github.com/ansible/ansible/issues/17330

abadger commented:
Occurring on:
- paramiko (2.0.2)
- OS Version: CentOS Linux release 7.2.1511 (Core)
- ansible 2.1.1.0

abadger commented:
Although filtering out None shouldn't hurt, we're not sure how None is getting into that list... It would be good for us to figure out how that's happening in case something deeper is going wrong.  Can you give us more information on how to fix this?

abadger commented on lib/ansible/plugins/connection/paramiko_ssh.py:
```diff
@@ -148,7 +148,7 @@ def _parse_proxy_command(self, port=22):
         ]
         if ssh_args is not None:
-            args = self._split_ssh_args(' '.join(ssh_args))
+            args = self._split_ssh_args(' '.join(filter(None,ssh_args)))
```
Note -- rather than using filter, it's more pythonic to use a generator expression or list comprehension like this:

``` python
args = self._split_ssh_args(' '.join(a for a in ssh_args if a is not None))
```

It can also be quicker if you are dealing with a large list (not the case here).

bcoca replied:
I would argue that a None here means there is a bug on the caller, this would obscure that bug, not fix it.

bcoca replied:
if we want to make the 'error nicer', we can capture the none-type exception and reissue and AnsibleError with friendlier text explaining why this happens.

abadger commented:
Sorry -- I didn't mean how to fix this.  I meant, can you give us more information on how to reproduce this.

veereshkhanorkar commented:
Let me share my code with you. I also had the same thoughts as you.

On Fri 9 Sep, 2016, 11:06 PM Toshio Kuratomi, notifications@github.com
wrote:

> Sorry -- I didn't mean how to fix this. I meant, can you give us more
> information on how to reproduce this.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/ansible/ansible/pull/17332#issuecomment-245983904,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AUYSK-qv0U-o6YzG6oKaTomCrBJTzBuIks5qoZkkgaJpZM4Jx9R-
> .

veereshkhanorkar commented:
The entry is via below:

<code>
  logger = options.logger
  tmpHome = os.path.join(options.dir, 'tmpHome')
  logger.info("Creating SSH Tmp home: %s", tmpHome)
  ssh = SSHInfra(tmpHome)
  ssh.createSSHFolder()
  runner = Runner(
      cwd = tmpHome,
      verbosity=5
  )
  runner.setBecome()
  playbooks = {}

  logger.info("Ansible command line: %s", options.ansible)
  for a in (b.split('=') for b in options.ansible.split(',')):
    playbooks[a[0]] = a[1]

  for hostgroup in playbooks:
    logger.info("Hostgroup: %s", hostgroup)
    logger.info("Hostgroup: %s, User: %s, Password: %s", hostgroup,
options.auth[hostgroup].user, options.auth[hostgroup].password)
    port = 22
    connection = 'paramiko'
    if 'WINDOW' in hostgroup:
      port = 5986
      connection = 'winrm'
    runner.addHostGroups(hostgroup,
options.auth[hostgroup].user,          options.auth[hostgroup].password,
port,    connection,     ['192.168.7.7', '192.168.7.2'])
    logger.info("Running %s against %s", playbooks[hostgroup], hostgroup)
    runner.setPlayBook(hostgroup, playbooks[hostgroup])
    run_data = {
      'hostgroup': hostgroup
    }
    stats = runner.run(runData = run_data, delete=False)
</code>

On Sat, Sep 10, 2016 at 9:05 AM, Khanorkar, Veeresh veeresh@khanorkar.com
wrote:

> Let me share my code with you. I also had the same thoughts as you.
> 
> On Fri 9 Sep, 2016, 11:06 PM Toshio Kuratomi, notifications@github.com
> wrote:
> 
> > Sorry -- I didn't mean how to fix this. I meant, can you give us more
> > information on how to reproduce this.
> > 
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > https://github.com/ansible/ansible/pull/17332#issuecomment-245983904,
> > or mute the thread
> > https://github.com/notifications/unsubscribe-auth/AUYSK-qv0U-o6YzG6oKaTomCrBJTzBuIks5qoZkkgaJpZM4Jx9R-
> > .

abadger commented:
Hmm... so you're not using the ansible CLi but trying to call this from python?  Perhaps you aren't initializing a value properly?  (Also, at least the names you're using [runner and such] make it look like you're tryign to use this from 1.x code, not from 2.x code?)

veereshkhanorkar commented:
Nope it's 2.0 code.

On Tue 13 Sep, 2016, 12:11 PM Toshio Kuratomi, notifications@github.com
wrote:

> Hmm... so you're not using the ansible CLi but trying to call this from
> python? Perhaps you aren't initializing a value properly? (Also, at least
> the names you're using [runner and such] make it look like you're tryign to
> use this from 1.x code, not from 2.x code?)
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/ansible/ansible/pull/17332#issuecomment-246589496,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AUYSKwwDXwl2RxyzQ2pYM2MradH6NC9Uks5qpkWagaJpZM4Jx9R-
> .

veereshkhanorkar commented:
Well, that's the right approach. Definitely, it will be obscured, the way I
have mentioned.

On Mon 19 Sep, 2016, 9:40 PM Brian Coca, notifications@github.com wrote:

> ## _@bcoca_ commented on this pull request.
> 
> In lib/ansible/plugins/connection/paramiko_ssh.py
> https://github.com/ansible/ansible/pull/17332:
> 
> > @@ -148,7 +148,7 @@ def _parse_proxy_command(self, port=22):
> >              getattr(self._play_context, 'ssh_args', ''),
> >          ]
> >          if ssh_args is not None:
> > -            args = self._split_ssh_args(' '.join(ssh_args))
> > -            args = self._split_ssh_args(' '.join(filter(None,ssh_args)))
> 
> if we want to make the 'error nicer', we can capture the none-type
> exception and reissue and AnsibleError with friendlier text explaining why
> this happens.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/ansible/ansible/pull/17332, or mute the thread
> https://github.com/notifications/unsubscribe-auth/AUYSKwy5imhoDJscmlw7mEu-cdroC_fMks5qrrPlgaJpZM4Jx9R-
> .

abadger commented:
Okay, so it seems like we decided this is a problem with the caller and not the connection code, is that right?  I'll close this next week unless someone thinks there is something we want to do.
